test-runtime/build.rs

```diff
@@ -0,0 +1,152 @@
+        }
+    };
+
+    // Download metadata from binary; retry until successful, or a limit is hit.
+    let res: serde_json::Value = {
+        const MAX_RETRIES: usize = 20;
+        let mut retries = 0;
+        let mut wait_secs = 1;
+        loop {
+            if retries >= MAX_RETRIES {
+                panic!("Cannot connect to substrate node after {} retries", retries);
+            }
+            let res = reqwest::blocking::Client::new()
+                .post(format!("http://localhost:{}", port))
+                .json(&json!({
+                   "id": 1,
+                   "jsonrpc": "2.0",
+                   "method": "state_getMetadata",
+                }))
+                .send();
+            match res {
+                Ok(res) if res.status().is_success() => {
+                    let _ = cmd.kill();
+                    break res
+                        .json()
+                        .expect("valid JSON response from substrate node expected")
+                }
+                _ => {
+                    thread::sleep(time::Duration::from_secs(wait_secs));
+                    retries += 1;
+                    wait_secs += 1;
+                }
+            };
+        }
+    };
+    let metadata_hex = res["result"]
+        .as_str()
+        .expect("Metadata should be returned as a string of hex encoded SCALE bytes");
+    let metadata_bytes = hex::decode(&metadata_hex.trim_start_matches("0x")).unwrap();
+
+    // Save metadata to a file:
+    let out_dir = env::var_os("OUT_DIR").unwrap();
+    let metadata_path = Path::new(&out_dir).join("metadata.scale");
+    fs::write(&metadata_path, metadata_bytes).expect("Couldn't write metadata output");
+
+    // Write out our expression to generate the runtime API to a file. Ideally, we'd just write this code
+    // in lib.rs, but we must pass a string literal (and not `concat!(..)`) as an arg to runtime_metadata_path,
+    // and so we need to spit it out here and include it verbatim instead.
+    let runtime_api_contents = format!(
+        "
+        #[subxt::subxt(
+            runtime_metadata_path = \"{}\",
+            generated_type_derives = \"Debug, Eq, PartialEq\"
+        )]
+        pub mod node_runtime {{
+            #[subxt(substitute_type = \"sp_arithmetic::per_things::Perbill\")]
+            use sp_runtime::Perbill;
+        }}
+    ",
+        metadata_path
+            .to_str()
+            .expect("Path to metadata should be stringifiable")
+    );
+    let runtime_path = Path::new(&out_dir).join("runtime.rs");
+    fs::write(&runtime_path, runtime_api_contents)
+        .expect("Couldn't write runtime rust output");
+
+    // Re-build if we point to a different substrate binary:
+    println!("cargo:rerun-if-env-changed={}", SUBSTRATE_BIN_ENV_VAR);
+    // Re-build if this file changes:
+    println!("cargo:rerun-if-changed=build.rs");
+}
+
+/// Returns the next open port, or None if no port found in range.
+fn next_open_port() -> Option<u16> {
```

Comment:
I wonder whether we could extract `TestNodeProcess` and other test utils to their own crate. Not only could we reuse the node spawning logic, but we could publish that crate for use by other projects for testing their nodes. I would certainly use it (this code is originally copied from `cargo-contract`).

Just an idea, don't need to do it here.

Reply:
I had a very similar thought doing this! I didn't really fancy the amount of churn in this PR, but I can imagine it would be useful!

Reply:
Will be worth doing, separate PR probably better - must remember remove the duplication here.

Reply:
why aren't we just binding to `0.0.0.0:0` instead and letting the OS find the next free port?

Reply:
That is copied from code I wrote in haste, so most likely not the best way to do it. If that works then let's do it.

Reply:
I do like the `0.0.0.0:0` approach normally; the complexity is that you then need to read the actual port used back to connect to the service, so I think I'd need to monitor the stdout (unless you have a better idea?).

In the other tests, we need to grab 3 ports by the looks of things, so it is a little trickier!

Reply:
But yes, in this build script, it's probably a little more robust to scan for a port than rely on any specific stdout from substrate on the matter (but if there's a better way to see which port was used, that'd be great!).